src/Formik.tsx

```diff
@@ -418,235 +314,415 @@ export class Formik<Values = FormikValues> extends React.Component<
+        const combinedErrors = deepmerge.all<FormikErrors<Values>>(
+          [fieldErrors, schemaErrors, validateErrors],
+          { arrayMerge }
+        );
+        if (!isEqual(state.errors, combinedErrors)) {
+          dispatch({ type: 'SET_ERRORS', payload: combinedErrors });
+        }
+        dispatch({ type: 'SET_ISVALIDATING', payload: false });
+        return combinedErrors;
+      });
+    } else {
+      return Promise.resolve({});
+    }
+  }
 
-      if (isPromise(maybePromisedOnReset)) {
-        (maybePromisedOnReset as Promise<any>).then(this.resetForm);
-      } else {
-        this.resetForm();
-      }
+  function setFormikState(
+    stateOrCb:
+      | FormikState<Values>
+      | ((state: FormikState<Values>) => FormikState<Values>)
+  ): void {
+    if (isFunction(stateOrCb)) {
+      dispatch({ type: 'SET_FORMIK_STATE', payload: stateOrCb(state) });
     } else {
-      this.resetForm();
+      dispatch({ type: 'SET_FORMIK_STATE', payload: stateOrCb });
     }
-  };
+  }
 
-  setFormikState = (s: any, callback?: (() => void)) =>
-    this.setState(s, callback);
+  function setStatus(status: any) {
+    dispatch({ type: 'SET_STATUS', payload: status });
+  }
 
-  validateForm = (values: Values) => {
-    this.setState({ isValidating: true });
-    return this.runValidations(values).then(errors => {
-      this.setState({ isValidating: false });
-      return errors;
+  function setSubmitting(isSubmitting: boolean) {
+    dispatch({ type: 'SET_ISSUBMITTING', payload: isSubmitting });
+  }
+
+  function submitForm() {
+    dispatch({ type: 'SUBMIT_ATTEMPT' });
+    return validateForm().then((combinedErrors: FormikErrors<Values>) => {
+      const isActuallyValid = Object.keys(combinedErrors).length === 0;
+      if (isActuallyValid) {
+        Promise.resolve(executeSubmit())
+          .then(() => {
+            if (isMounted.current) {
+              dispatch({ type: 'SUBMIT_SUCCESS' });
+            }
+          })
+          .catch(_errors => {
+            if (isMounted.current) {
+              dispatch({ type: 'SUBMIT_FAILURE' });
+            }
+          });
+      } else if (isMounted.current) {
+        // ^^^ Make sure Formik is still mounted before calling setState
+        dispatch({ type: 'SUBMIT_FAILURE' });
+      }
     });
-  };
+  }
 
-  getFormikActions = (): FormikActions<Values> => {
-    return {
-      resetForm: this.resetForm,
-      submitForm: this.submitForm,
-      validateForm: this.validateForm,
-      validateField: this.validateField,
-      setError: this.setError,
-      setErrors: this.setErrors,
-      setFieldError: this.setFieldError,
-      setFieldTouched: this.setFieldTouched,
-      setFieldValue: this.setFieldValue,
-      setStatus: this.setStatus,
-      setSubmitting: this.setSubmitting,
-      setTouched: this.setTouched,
-      setValues: this.setValues,
-      setFormikState: this.setFormikState,
+  function getFieldProps(
+    name: string,
+    type: string
```

Comment:
Couldn't `type` be optional here?
```suggestion
    type?: string
```